File: exercise/5/manacher.cpp
static const auto speedup =[](){std::ios::sync_with_stdio(false); std::cin.tie(NULL); return 0;}();
class Solution {
public:
    string longestPalindrome(string s) {
		int max_length = 0, pos = 0;
        string p_s;
		p_s.push_back('#');
		for (int i = 0; i < s.length(); i++) {
			p_s.push_back(s[i]);
			p_s.push_back('#');
		}
		vector<int> longest(p_s.length());
		int right = 0, center = 0;
		for (int i = 1; i < p_s.length(); i++) {
            if (i >= right) {
                longest[i] = 0;
            } else {
                longest[i] = min(longest[2 * center - i], right - i);
            }
            for (;i - longest[i] - 1 >= 0 && i + longest[i] + 1 < p_s.length() && p_s[i-longest[i]-1] == p_s[i+longest[i]+1]; longest[i]++);
            if (right < i + longest[i]) {
                right = i + longest[i];
                center = i;
            }
            if (max_length < longest[i]) {
                max_length = longest[i];
                pos = i;
            }
		}
		string result;
		for (int i = pos - max_length; i <= pos + max_length; i++) {
			if (p_s[i] != '#') {
				result.push_back(p_s[i]);
			}
		}
		return result;
    }
};

File: exercise/5/manacher.py
class Solution:
    def longestPalindrome(self, s: str) -> str:
        s_trans = '#'
        for i in range(0, len(s)):
            s_trans += s[i]
            s_trans += '#'
        length = [0 for _ in range(len(s_trans))]
        max_length = 0
        max_mid = 0
        center = 0
        right = 0
        for i in range(1, len(s_trans)):
            if i >= right:
                length[i] = 0
            else:
                length[i] = min(right - i, length[2 * center - i])
            while i - length[i] - 1 >= 0 and i + length[i] + 1 < len(s_trans):
                if s_trans[i - length[i] - 1] == s_trans[i + length[i] + 1]:
                    length[i] += 1
                else:
                    break
            if i + length[i] > right:
                right = i + length[i]
                center = i
            if length[i] > max_length:
                max_length = length[i]
                max_mid = i
        result = ''
        for i in range(max_mid - max_length, max_mid + max_length + 1):
            if s_trans[i] != '#':
                result += s_trans[i]
        return result
